### ISSUE TITLE
Compilation error: Generic subtype and its generic supertype

### ISSUE THREAD

**Issue by Ragmaanir** [crystal-lang/crystal#1411]
Hi, i get an error with this code:

```crystal
class ExpSpec < Minitest::Test

  abstract class Bar(T)
    abstract def data : T
  end

  class ExtraBar(T) < Bar(T)
    def initialize(@value : T)
    end

    def data : T
      @value
    end
  end

  def test_it
    b :: Bar(Int32)
    e = ExtraBar.new(12)

    b = e as Bar(Int32) # XXX

    p b.class
    p b.data
  end

end
```

```
  1) Error:
ExpSpec#test_it [:]:
Exception: can't cast to ExpSpec::Bar(Int32) because it was never instantiated
    *Exception#initialize<Exception, String, Nil>:Array(String) +14 [4329729966]
    *Exception::new<String>:Exception +113 [4329729921]
    *raise<String>:NoReturn +6 [4329684438]
    *ExpSpec#test_it<ExpSpec>:NoReturn +38 [4329852550]
```

When i remove `as Bar(Int32)` from the line marked with `XXX` then i get this compilation error:
```
type must be ExpSpec::Bar(Int32), not (ExpSpec::Bar(Int32) | ExpSpec::ExtraBar(Int32))
```

Since `ExtraBar(T)` is defined as `class ExtraBar(T) < Bar(T)` i would expect that a variable declared as `Bar(Int32)` (the superclass of `ExtraBar(Int32)`) could reference instances of the latter type. Why do i get these errors? Or am i missing something?

**Comment by jhass** [comment]
I think this boils down to a duplicate of #1297

**Comment by Ragmaanir** [comment]
@jhass yeah, i think you are right. i'm closing this one then.

**Comment by Ragmaanir** [comment]
Actually, im not sure anymore. In the mentioned ticket the problem is this:
```
Array(Subclass).is_a?(Array(Superclass))
```

While in this case it is:
```
Subclass(X).is_a?(Superclass(X))
```
WDYT?

**Comment by asterite** [comment]
@Ragmaanir Quick question: why do you need this line?

```crystal
e as Bar(Int32)
```

Also doing `b :: Bar(Int32)` is [unsafe](http://crystal-lang.org/docs/syntax_and_semantics/declare_var.html).

Casting types to supertypes in Crystal has no effect at all. Casting only makes sense when downcasting a type "X" to "Y" where "X < Y".

**Comment by Ragmaanir** [comment]
I want to make a variable with type `Superclass` to refer to an instance of type `Subclass`. If i dont add the `e as Bar(Int32)`, i get the compilation error mentioned in the first post.

If `b :: Bar(Int32)` is unsafe, how do i specify the type of a local variable?

**Comment by asterite** [comment]
This works:

```crystal
abstract class Bar(T)
  abstract def data : T
end

class ExtraBar(T) < Bar(T)
  def initialize(@value : T)
  end

  def data : T
    @value
  end
end

e = ExtraBar.new(12)
p e.data #=> 12
```

I still don't understand why you need to cast `e` to `Bar(Int32)`. The compiler knows that `e` is `ExtraBar`.

A different example:

```crystal
class Foo
  def method
    1
  end
end

class Bar < Foo
  def method
    2
  end
end

bar = Bar.new
foo = bar as Foo
puts foo.method #=> prints 2
```

That is, the `bar as Foo` has almost no effect. The compiler will now treat `foo` as ` Foo` or any of it's subclasses, but at runtime it's still a `Bar` .

As for defining the type of a local variable, you generally don't do that. You just let the compiler infer it based on what you assigned to it.

I know this is pretty hard to get used to if you come from a more classical typed language, but it works well in Ruby and also in Crystal.

However, it's still an issue that you can't cast `ExtraBar(Int32)` to `Bar(Int32)`, so I'll leave this issue open. This is mostly an issue because we distinguish between exact types T and virtual types T (T or any of its subclasses), but more and more I'm convinced that that distinction should go away.

**Comment by Ragmaanir** [comment]
The example is simplified. In my project its a little more complicated and involves the return type of methods. I am still trying to narrow it down to a minimal example (my current minimal example compiles and works, but my project doesnt), but my time is limited atm. 

Specifying the type of local variables and other things is very helpful IMO, especially when you get compilation errors. It would be helpful to be able to make sure that a specific expression has a specific type, or at least be able to see what the compiler inferred (in the presence of compilation errors). 

Since i am programming in both ruby and scala i know both worlds, but in crystal the typing is a bit tricky some times due to automatic inference of union types.

**Comment by jhass** [comment]
@Ragmaanir anymore feedback? I'd like to close this otherwise.

**Comment by Ragmaanir** [comment]
I'm back home now and finally have an internet connection again.

Here is another example that does not cast, it just specifies the return type of a method.

```crystal

class ExpSpec < Minitest::Test

  abstract class Bar(T)
    abstract def data : T
  end

  class ExtraBar < Bar(Int32)
    def initialize(@data : Int32)
    end

    getter :data
  end

  def create : Bar(Int32)
    ExtraBar.new(12)
  end

  def test_it
    b = create

    p b.class
    p b.data
  end

end
```

When compiling with crystal 0.9.0 i get:
```
type must be ExpSpec::Bar(Int32), not ExpSpec::ExtraBar

  def create : Bar(Int32)
```

Shouldnt it compile since ExtraBar is a subclass of Bar(Int32). Or am i missing something?

**Comment by asterite** [comment]
Yes, it should compile. These aren't working right now mostly because we don't use these annotations, but they should definitely work.

**Comment by asterite** [comment]
Closed in favor of #2665
